src/v/storage/segment_deduplication_utils.h

```diff
@@ -0,0 +1,49 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "model/fundamental.h"
+#include "seastarx.h"
+#include "storage/fwd.h"
+#include "storage/index_state.h"
+#include "storage/segment_set.h"
+
+namespace storage {
+using segment_list_t = fragmented_vector<segment_set::type>;
+
+// Adds the keys from the given compacted index reader to the map. Returns
+// true if the entire reader was successfully indexed.
+ss::future<bool> build_offset_map_for_segment(
+  const compaction_config& cfg, const segment& seg, key_offset_map& m);
+
+// Builds a map from key to latest offset from the last segment to the
+// earliest segment in 'segs'.
+//
+// Returns the start offset of the earliest segment that was fully indexed.
+// The resulting map may contain offsets below this point (i.e. the
+// preceding segment may have been partially indexed), but to fully
+// deduplicate the log, subsequent compactions should start below this
+// offset.
+//
+// Returns nullopt if there was a problem building the map.
+ss::future<std::optional<model::offset>> build_offset_map(
+  const compaction_config& cfg, const segment_set& segs, key_offset_map& m);
```

Comment:
nit: Looks like it returns nullopt or throws in the failure case. Maybe throw in all cases to simplify the interface?

Reply:
Done